Frameworks/CoreGraphics/DWriteWrapper.mm

```diff
@@ -0,0 +1,572 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+// Do not move this block, it has to come first for some reason
+#include <COMIncludes.h>
+#import <wrl/implements.h>
+#include <COMIncludes_End.h>
+
+#import <CoreGraphics/DWriteWrapper.h>
+#import <Starboard.h>
+
+#import <vector>
+
+using namespace std;
+using namespace Microsoft::WRL;
+
+static const wchar_t* TAG = L"_DWriteWrapper";
+static const wchar_t* c_defaultUserLanguage = L"en-us";
+
+/**
+ * Private helper method to return the user set default locale string.
+ *
+ * @return use set locale string as wstring.
+ */
+static wstring __GetUserDefaultLocaleName() {
+    wchar_t localeName[LOCALE_NAME_MAX_LENGTH];
+    int defaultLocaleSuccess = GetUserDefaultLocaleName(localeName, LOCALE_NAME_MAX_LENGTH);
+
+    // If the default locale is returned, find that locale name, otherwise use "en-us".
+    if (defaultLocaleSuccess) {
+        return wstring(localeName);
+    } else {
+        return wstring(c_defaultUserLanguage);
+    }
+}
+
+/**
+ * Helper method to convert IDWriteLocalizedStrings object to CFString object.
+ *
+ * @parameter localizedString IDWriteLocalizedStrings object to convert.
+ *
+ * @return CFString object.
+ */
+CFStringRef _CFStringFromLocalizedString(IDWriteLocalizedStrings* localizedString) {
+    if (localizedString == NULL) {
+        TraceError(TAG, L"The input parameter is invalid!");
+        return nil;
+    }
+
+    // Get the default locale for this user.
+    wstring localeName = __GetUserDefaultLocaleName();
+
+    uint32_t index = 0;
+    BOOL exists = false;
+
+    // If the default locale is returned, find that locale name, otherwise use "en-us".
+    THROW_IF_FAILED(localizedString->FindLocaleName(localeName.c_str(), &index, &exists));
+    if (!exists) {
+        THROW_IF_FAILED(localizedString->FindLocaleName(c_defaultUserLanguage, &index, &exists));
+    }
+
+    // If the specified locale doesn't exist, select the first on the list.
+    if (!exists) {
+        index = 0;
+    }
+
+    // Get the string length.
+    uint32_t length = 0;
+    THROW_IF_FAILED(localizedString->GetStringLength(index, &length));
+
+    // Get the string.
+    vector<wchar_t> wcharString = std::vector<wchar_t>(length + 1);
+    THROW_IF_FAILED(localizedString->GetString(index, wcharString.data(), length + 1));
```

Comment:
getting just length causes a throw

---
In reply to: [84335143](https://github.com/Microsoft/WinObjC/pull/1145#discussion_r84335143) [](ancestors = 84335143)